packages/protocol/contracts/layer1/preconf/iface/IPreconfSlasher.sol

```diff
@@ -0,0 +1,144 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.24;
+
+import { ITaikoInbox } from "src/layer1/based/ITaikoInbox.sol";
+import { ISlasher } from "@eth-fabric/urc/ISlasher.sol";
+
+/// @title IPreconfSlasher
+/// @custom:security-contact security@taiko.xyz
+interface IPreconfSlasher is ISlasher {
+    // Byte-encoded and used as `ISlasher.Commitment.payload`.
+    struct CommitmentPayload {
+        // Taiko specific DS
+        bytes32 domainSeparator;
+        // Chainid of Taiko
+        uint256 chainId;
+        // Timestamp of the L1 slot in which the preconfer is the proposer.
+        // In the case of falback preconfer, this will be the timestamp of the last slot of the
+        // epoch.
+        uint256 l1ProposalSlotTimestamp;
+        // Height of the L1 anchor block
+        uint256 anchorId;
+        // Blockhash of the L1 anchor block
+        bytes32 anchorHash;
+        // ID of the batch that will be containing this block in TaikoInbox
+        uint256 batchId;
+        // Hash of the header of the preconfirmed block
+        bytes32 blockHash;
+        // `true` if this preconfer is not going to deliver anymore
+        // preconfirmations after this block
+        bool eop; // End-Of-Preconf flag
+    }
+
+    // Evidence to prove preconfirmation violation of a block at a certain height X
+    struct Evidence {
```

Comment:
May I suggest we define a more generic proof struct like this:

```
struct Evidence {
ViolationType violationType;
bytes evidenceData;
}
```
Then depends on the violationType, we decode evidenceData into different sub evidence struct.

Reply:
Or maybe we don't need such a generic strut at all, in the slash function, we can just take the first byte and use it to check the violationType, then we do something like:

```
if (_evidence[:1] == uint(ViolationType.xyz)) {
   EvidenceXyz memory evidence = abi.decode(_evidence[1:], (EvidenceXyz));
   
   // route to a library to minimize the code of this contract so this contract?
   LibEvidenceXyz.slash(evidence) 
 } else ...
 
```